### PR TITLE
Update test utils for transformers 4.40

### PR DESCRIPTION
**Pull request by malkomes**
# What does this PR do?

- Updates `test_utils` and `test_framework_agnostic` according to transformers `4.40` tests.
- Add the possibility of `decoder_start_token_id` to be a `list` when `token_idx is not None`.
- Raises an `AssertionError` if using `inputs_embeds` and `token_idx` since `input_ids` will be empty and this code path crashes.
- Raises an `ValueError` if the user requests scores when using `static_shapes`, no scores are computed in this case.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/habana/transformers/generation/utils.py [hunk 1]
```diff
@@ -952,6 +954,10 @@ def generate(
             input_ids_length=input_ids_length,
             has_token_idx="token_idx" in model_kwargs,
         )
+        if model_input_name == "inputs_embeds" and "token_idx" in model_kwargs:
+            # Need to make sure that generation_config.max_length is correct in this code path
+            # otherwise it will crash withs self._validate_generated_length() below
+            assert input_ids.numel() > 0, "optimum-habana does not support inputs_embeds yet"
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:960]: The issue here is that `inputs_embeds` are not supported with static shapes right?

**Review by regisss — COMMENTED** [review]
Please run `make style`

File: optimum/habana/transformers/generation/utils.py (hunk 1 above)
Comment (malkomes) [reply optimum/habana/transformers/generation/utils.py:960]: correct. Reading it now, I think a `ValueError` error would make more sense. The fact that `input_ids` at this point empty it's just a by product of the two conditions of the `if`.

**Comment by yafshar** [comment]
@malkomes, please re-base this PR with the main. I will check it as soon you are done.

File: tests/transformers/tests/generation/test_utils.py [hunk 2]
```diff
@@ -45,41 +52,31 @@
...
-        TopKLogitsWarper,
-        TopPLogitsWarper,
     )
     from transformers.generation.candidate_generator import AssistedCandidateGenerator, CandidateGenerator
-    from transformers.generation.streamers import BaseStreamer
-
-torch_device = "hpu"
-adapt_transformers_to_gaudi()
+    from transformers.generation.utils import _speculative_sampling
+    from transformers.streamers import BaseStreamer
```
Comment (yafshar) [inline tests/transformers/tests/generation/test_utils.py:79]:
```suggestion
    from transformers.generation.streamers import BaseStreamer
```
@malkomes fix the import!

Comment (yafshar) [reply tests/transformers/tests/generation/test_utils.py:79]: if you re-base to main this is already there

Comment (malkomes) [reply tests/transformers/tests/generation/test_utils.py:79]: I'm on it

**Comment by malkomes** [comment]
Please, note that this PR does upgrade the tests to 4.40 but it doesn't fix all the issues. Specifically, I'm expecting the following tests to fail: 

| Test | Comment |
|------|---------|
| GenerationIntegrationTests::test_beam_search_low_memory                       | RuntimeError: Incompatible input shapes, broadcast not possible. Tensor1 Size: 42 1 12 1 Tensor2  |
| GenerationIntegrationTests::test_custom_stopping_criteria                     | AssertionError: Lists differ: [1, 22] != [1, 20]  |
| GenerationIntegrationTests::test_eos_token_id_int_and_list_beam_search        | AssertionError: False is not true  |
| GenerationIntegrationTests::test_eos_token_id_int_and_list_contrastive_search | AssertionError: False is not true  |
| GenerationIntegrationTests::test_eos_token_id_int_and_list_greedy_search      | AssertionError: False is not true  |
| GenerationIntegrationTests::test_generate_vision2text_conditioning            | RuntimeError: Incompatible input shapes, broadcast not possible. Tensor1 Size: 32 5 2 Tensor2 Size: 32 6 2  |
| GenerationIntegrationTests::test_max_new_tokens_encoder_decoder               | RuntimeError: Graph compile failed. synStatus=synStatus 26 [Generic failure].  |
| GenerationIntegrationTests::test_stop_sequence_stopping_criteria              | TypeError: unsupported operand type(s) for -: 'NoneType' and 'int'  |

Since this PR is already long and I'm not 100% sure what's the best fix for some of these issues, I propose and I would be happy to further investigate them independently in different PRs.

**Comment by yafshar** [comment]
@malkomes I am checking the issues. Can you share which tests have you done!

**Comment by malkomes** [comment]
Not sure if I understand your ask. I can give you a list of what we have in `main` vs `branch`, is that what you want? 

Here's a short summary. `test_utils` in main collects 48 tests where 22 failed, 17 passed, 9 skipped. 
This branch has 59 tests, of which 8 failed, 26 passed, 9 skipped, 16 xfailed. I've marked fails the tests which raise `NotImplementedError` or have a feature that is not supported yet. 

I share more details on the GS-166 ticket.

File: optimum/habana/transformers/generation/utils.py [hunk 3]
```diff
@@ -193,24 +193,26 @@ def _prepare_decoder_input_ids_for_generation(
         decoder_start_token_id = self._get_decoder_start_token_id(decoder_start_token_id, bos_token_id)
         if device is None:
             device = self.device
-        if token_idx is None:
```
Comment (ssarkar2) [inline optimum/habana/transformers/generation/utils.py:196]: @malkomes : I am not sure or cant remember why this check was there. I see you have removed it, can you comment why this is removed?

Comment (malkomes) [reply optimum/habana/transformers/generation/utils.py:196]: sure @ssarkar2 , here I just re-organize the logic and it is kinda of subtle. 

Before it was something like:

 - if not static shape:
   - is batch?
     - validate batch
     - re-org batch shape for `decoder_input_ids_start` 
   - (else) not a batch
     - `decoder_input_ids_start` <- ones * `decoder_start_token_id`
 - (else) using static shape:
   - max_length 
   - `decoder_input_ids_start` <- ones (with max_length) * `decoder_start_token_id`

multiple values for `decoder_start_token_id` wouldn't work when using static shape.

now the code reads something like:

- is batch?
   - validate batch
   - re-org batch shape for `decoder_input_ids_start` 
- (else) not a batch
    - `decoder_input_ids_start` <- ones * `decoder_start_token_id`
- if is static shape:
   - max_length
   - (the `decoder_input_ids_start` now has the expected format for both cases above, so we just need to apply the "static shape" logic)
   - `decoder_input_ids_start` <- ones (with max_length) * `decoder_input_ids_start`

Let me know if that makes sense.

**Comment by malkomes** [comment]
Recent changes to main branch might have made this PR obsolete.

**Comment by yafshar** [comment]
@malkomes thanks for your work and sorry for being late in reviewing this PR. 
- Can you start this PR again? 
- If it is obsolete, can you close this one and open a new one targeted for the main? 
Please ping me as soon as you start and I will make sure to review and update with your progress. Thanks

**Comment by malkomes** [comment]
sure, @yafshar . I'm back from PTO and the `huggingface:main` has changed a little bit. I'm going to go with the second option of closing this one and pinging you on the next PR, I think that's going to be cleaner. Thanks

**Pull request closed without merging**